File: sdk/python/src/protomcp/resource.py
import json
from dataclasses import dataclass, field
from typing import Callable, Optional

_resource_registry: list["ResourceDef"] = []
_resource_template_registry: list["ResourceTemplateDef"] = []

@dataclass
class ResourceDef:
    uri: str
    name: str
    description: str
    handler: Callable  # (uri: str) -> ResourceContent or list[ResourceContent]
    mime_type: str = ""
    size: int = 0

@dataclass
class ResourceContent:
    uri: str
    text: str = ""
    blob: bytes = b""
    mime_type: str = ""

@dataclass
class ResourceTemplateDef:
    uri_template: str
    name: str
    description: str
    handler: Callable  # (uri: str) -> ResourceContent or list[ResourceContent]
    mime_type: str = ""

def resource(uri: str, description: str, name: str = "", mime_type: str = ""):
    """Decorator to register a resource handler."""
    def decorator(func: Callable) -> Callable:
        _resource_registry.append(ResourceDef(
            uri=uri,
            name=name or func.__name__,
            description=description,
            handler=func,
            mime_type=mime_type,
        ))
        return func
    return decorator

def resource_template(uri_template: str, description: str, name: str = "", mime_type: str = ""):
    """Decorator to register a resource template handler."""
    def decorator(func: Callable) -> Callable:
        _resource_template_registry.append(ResourceTemplateDef(
            uri_template=uri_template,
            name=name or func.__name__,
            description=description,
            handler=func,
            mime_type=mime_type,
        ))
        return func
    return decorator

def get_registered_resources() -> list[ResourceDef]:
    return list(_resource_registry)

def get_registered_resource_templates() -> list[ResourceTemplateDef]:
    return list(_resource_template_registry)

File: sdk/go/protomcp/resource.go
package protomcp

// ResourceContent represents a single content item returned by a resource handler.
type ResourceContent struct {
	URI      string
	MimeType string
	Text     string
	Blob     []byte
}

// ResourceDef represents a registered static resource.
type ResourceDef struct {
	URI         string
	Name        string
	Description string
	MimeType    string
	Size        int64
	HandlerFn   func() []ResourceContent
}

// ResourceTemplateDef represents a registered resource template.
type ResourceTemplateDef struct {
	URITemplate string
	Name        string
	Description string
	MimeType    string
	HandlerFn   func(uri string) []ResourceContent
}

var resourceRegistry []ResourceDef
var resourceTemplateRegistry []ResourceTemplateDef

// RegisterResource adds a static resource to the global registry.
func RegisterResource(def ResourceDef) {
	resourceRegistry = append(resourceRegistry, def)
}

// RegisterResourceTemplate adds a resource template to the global registry.
func RegisterResourceTemplate(def ResourceTemplateDef) {
	resourceTemplateRegistry = append(resourceTemplateRegistry, def)
}

// GetRegisteredResources returns a copy of all registered resources.
func GetRegisteredResources() []ResourceDef {
	return append([]ResourceDef{}, resourceRegistry...)
}

// GetRegisteredResourceTemplates returns a copy of all registered resource templates.
func GetRegisteredResourceTemplates() []ResourceTemplateDef {
	return append([]ResourceTemplateDef{}, resourceTemplateRegistry...)
}

// ClearResourceRegistry resets the resource registries.
func ClearResourceRegistry() {
	resourceRegistry = nil
	resourceTemplateRegistry = nil
}

// Resource is a convenience function to register a static resource.
func Resource(uri, description string, handler func() []ResourceContent) {
	RegisterResource(ResourceDef{
		URI:         uri,
		Name:        uri,
		Description: description,
		HandlerFn:   handler,
	})
}

// TextResource registers a static resource that returns a single text content.
func TextResource(uri, description, text string) {
	Resource(uri, description, func() []ResourceContent {
		return []ResourceContent{{URI: uri, Text: text}}
	})
}

// ResourceTemplate is a convenience function to register a resource template.
func ResourceTemplate(uriTemplate, description string, handler func(uri string) []ResourceContent) {
	RegisterResourceTemplate(ResourceTemplateDef{
		URITemplate: uriTemplate,
		Name:        uriTemplate,
		Description: description,
		HandlerFn:   handler,
	})
}
